sys/net/gnrc/link_layer/lorawan/gnrc_lorawan_mlme.c

```diff
@@ -0,0 +1,328 @@
+}
+
+void gnrc_lorawan_mlme_request(gnrc_lorawan_t *mac, const mlme_request_t *mlme_request,
+                               mlme_confirm_t *mlme_confirm)
+{
+    switch (mlme_request->type) {
+        case MLME_JOIN:
+            if(mac->mlme.activation != MLME_ACTIVATION_NONE) {
+                mlme_confirm->status = -EINVAL;
+                break;
+            }
+            if (!gnrc_lorawan_mac_acquire(mac)) {
+                mlme_confirm->status = -EBUSY;
+                break;
+            }
+
+            if (mac->mlme.backoff_budget < 0) {
+                mlme_confirm->status = -EDQUOT;
+                break;
+            }
+            memcpy(mac->appskey, mlme_request->join.appkey, LORAMAC_APPKEY_LEN);
+            mlme_confirm->status = gnrc_lorawan_send_join_request(mac, mlme_request->join.deveui,
+                                                                  mlme_request->join.appeui, mlme_request->join.appkey, mlme_request->join.dr);
+            break;
+        case MLME_LINK_CHECK:
+            mac->mlme.pending_mlme_opts |= GNRC_LORAWAN_MLME_OPTS_LINK_CHECK_REQ;
+            mlme_confirm->status = GNRC_LORAWAN_REQ_STATUS_DEFERRED;
+            break;
+        case MLME_SET:
+            _mlme_set(mac, mlme_request, mlme_confirm);
+            break;
+        case MLME_GET:
+            _mlme_get(mac, mlme_request, mlme_confirm);
+            break;
+        case MLME_RESET:
+            gnrc_lorawan_reset(mac);
+            mlme_confirm->status = GNRC_LORAWAN_REQ_STATUS_SUCCESS;
+            break;
+        default:
+            break;
+    }
+}
+
+int _fopts_mlme_link_check_req(lorawan_buffer_t *buf)
+{
+    if (buf) {
+        assert(buf->index + GNRC_LORAWAN_CID_SIZE <= buf->size);
+        buf->data[buf->index++] = GNRC_LORAWAN_CID_LINK_CHECK_ANS;
+    }
+
+    return GNRC_LORAWAN_CID_SIZE;
+}
+
+static void _mlme_link_check_ans(gnrc_lorawan_t *mac, uint8_t *p)
+{
+    mlme_confirm_t *mlme_confirm = gnrc_lorawan_mlme_allocate(mac);
+    mlme_confirm->link_req.margin = p[0];
+    mlme_confirm->link_req.num_gateways = p[1];
+
+    mlme_confirm->type = MLME_LINK_CHECK;
+    mlme_confirm->status = GNRC_LORAWAN_REQ_STATUS_SUCCESS;
+    mac->netdev.event_callback(&mac->netdev, NETDEV_EVENT_MLME_CONFIRM);
+
+    mac->mlme.pending_mlme_opts &= ~GNRC_LORAWAN_MLME_OPTS_LINK_CHECK_REQ;
+}
+
+void gnrc_lorawan_process_fopts(gnrc_lorawan_t *mac, uint8_t *fopts, size_t size)
+{
+    if (!fopts || !size) {
+        return;
+    }
+
+    uint8_t ret = 0;
+    void (*cb)(gnrc_lorawan_t*, uint8_t *p) = NULL;
+
+    for(uint8_t pos = 0; pos < size; pos += ret) {
+        switch (fopts[pos]) {
+            case GNRC_LORAWAN_CID_LINK_CHECK_ANS:
+                ret += GNRC_LORAWAN_FOPT_LINK_CHECK_ANS_SIZE;
+                cb = _mlme_link_check_ans;
+                break;
+            default:
+                return;
+        }
+
+        if(pos + ret > size) {
+            return;
+        }
+
+        cb(mac, fopts + pos + 1);
```

Comment:
last one: use `&fopts[pos + 1]` or (even better (to me)) `&fopts[pos]` and then use `p[1]` and `p[2]` in the other function. The latter would also allow for later changing this to a (nice) struct casting based solution

```
typedef struct {
    uint8_t type;
    uint8_t margin;
    uint8_t ngtws;
} mlme_link_check_ans_t;

mlme_link_check_ans_t *mlca = (mlme_link_check_ans_t *)p;

mlme_confirm->link_req.margin = mlca->margin;
mlme_confirm->link_req.num_gateways = mlca->ngtws;
```

or something like that.

I'd prefer the bracket notation bc its used in the switch, too.

Reply:
done :) (https://github.com/RIOT-OS/RIOT/pull/11022/commits/407d365d7009c8d494f749f568ef2cd70f9f7d0e)